File: packages/sdk-typescript/src/services/knowledgeBase.ts
/**
 * 知识库管理服务
 */

import type { 
  KBJobCreate, 
  KBJobResponse, 
  KBJobStatus 
} from '../types';

export class KnowledgeBaseService {
  private client: any; // KnowhereClient

  constructor(client: any) {
    this.client = client;
  }

  /**
   * 创建知识库管理任务
   */
  async createJob(params: KBJobCreate): Promise<KBJobResponse> {
    const response = await this.client.request('/v1/kb/jobs', {
      method: 'POST',
      body: JSON.stringify(params),
    });

    return response.data;
  }

  /**
   * 获取任务状态
   */
  async getJobStatus(jobId: string): Promise<KBJobStatus> {
    const response = await this.client.request(`/v1/kb/jobs/${jobId}`);
    return response.data;
  }

  /**
   * 下载处理结果
   */
  async downloadResult(jobId: string): Promise<any> {
    const response = await this.client.request(`/v1/kb/jobs/${jobId}/download`);
    return response.data;
  }

  /**
   * 等待任务完成
   */
  async waitForCompletion(
    jobId: string,
    timeout: number = 3600,
    pollInterval: number = 5000
  ): Promise<KBJobStatus> {
    const startTime = Date.now();

    while (Date.now() - startTime < timeout * 1000) {
      const status = await this.getJobStatus(jobId);

      if (['completed', 'failed'].includes(status.status)) {
        return status;
      }

      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }

    throw new Error(`任务 ${jobId} 在 ${timeout} 秒内未完成`);
  }
}

File: packages/sdk-python/knowhere_sdk/services/knowledge_base.py
"""
知识库管理服务
"""

from typing import Dict, Any, Optional
from ..client import KnowhereClient
from ..types import KBJobCreate, KBJobResponse, KBJobStatus


class KnowledgeBaseService:
    """知识库管理服务"""
    
    def __init__(self, client: KnowhereClient):
        self.client = client
    
    async def create_job(
        self,
        file_url: str,
        webhook_url: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> KBJobResponse:
        """
        创建知识库管理任务
        
        Args:
            file_url: 文档文件URL
            webhook_url: Webhook通知URL（可选）
            metadata: 额外元数据（可选）
            
        Returns:
            KBJobResponse: 任务创建响应
        """
        data = {
            "file_url": file_url,
            "webhook_url": webhook_url,
            "metadata": metadata or {}
        }
        
        response = await self.client._request(
            "POST",
            "/v1/kb/jobs",
            data=data
        )
        
        return KBJobResponse(**response.data)
    
    async def get_job_status(self, job_id: str) -> KBJobStatus:
        """
        获取任务状态
        
        Args:
            job_id: 任务ID
            
        Returns:
            KBJobStatus: 任务状态
        """
        response = await self.client._request(
            "GET",
            f"/v1/kb/jobs/{job_id}"
        )
        
        return KBJobStatus(**response.data)
    
    async def download_result(self, job_id: str) -> Dict[str, Any]:
        """
        下载处理结果
        
        Args:
            job_id: 任务ID
            
        Returns:
            Dict[str, Any]: 处理结果数据
        """
        response = await self.client._request(
            "GET",
            f"/v1/kb/jobs/{job_id}/download"
        )
        
        return response.data
    
    async def wait_for_completion(
        self,
        job_id: str,
        timeout: int = 3600,
        poll_interval: int = 5
    ) -> KBJobStatus:
        """
        等待任务完成
        
        Args:
            job_id: 任务ID
            timeout: 超时时间（秒）
            poll_interval: 轮询间隔（秒）
            
        Returns:
            KBJobStatus: 最终任务状态
        """
        import asyncio
        import time
        
        start_time = time.time()
        
        while time.time() - start_time < timeout:
            status = await self.get_job_status(job_id)
            
            if status.status in ['completed', 'failed']:
                return status
            
            await asyncio.sleep(poll_interval)
        
        raise TimeoutError(f"任务 {job_id} 在 {timeout} 秒内未完成")
